test/solvers/advanced/forward_backward_test.py

```diff
@@ -0,0 +1,227 @@
+
+    # Check that the algorithm runs. With the above operators, the algorithm
+    # returns the input.
+    x0 = example_element(space1)
+    x = x0.copy()
+    niter = 3
+
+    forward_backward_pd(x, prox_f, prox_cc_g, lin_ops, grad_h, tau=1.0,
+                        sigma=[1.0, 1.0], niter=niter)
+
+    assert x == x0
+
+    # Testing that sizes needs to agree:
+    # Too few sigma_i:s
+    with pytest.raises(ValueError):
+        forward_backward_pd(x, prox_f, prox_cc_g, lin_ops, grad_h, tau=1.0,
+                            sigma=[1.0], niter=niter)
+
+    # Too many operators
+    prox_cc_g_too_many = [odl.solvers.proximal_zero(space1),
+                          odl.solvers.proximal_zero(space1),
+                          odl.solvers.proximal_zero(space1)]
+    with pytest.raises(ValueError):
+        forward_backward_pd(x, prox_f, prox_cc_g_too_many, lin_ops, grad_h,
+                            tau=1.0, sigma=[1.0, 1.0], niter=niter)
+
+    # Test for correct space
+    space2 = odl.uniform_discr(1, 2, 10)
+    x = example_element(space2)
+    with pytest.raises(ValueError):
+        forward_backward_pd(x, prox_f, prox_cc_g, lin_ops, grad_h, tau=1.0,
+                            sigma=[1.0, 1.0], niter=niter)
+
+
+def test_forward_backward_basic():
+    """Test for the forward-backward solver by minimizing ||x||_2^2.
+
+    The general problem is of the form
+
+        ``min_x f(x) + sum_i g_i(L_i x) + h(x)``
+
+    and here we take f(x) = g(x) = 0, h(x) = ||x||_2^2 and L is the
+    zero-operator.
+    """
+
+    space = odl.rn(10)
+
+    lin_ops = [odl.ZeroOperator(space)]
+    prox_cc_g = [odl.solvers.proximal_cconj(odl.solvers.proximal_zero(space))]
+    prox_f = odl.solvers.proximal_zero(space)
+    grad_h = 2.0 * odl.IdentityOperator(space)  # Gradient of two-norm square
+
+    x = example_element(space)
+    x_global_min = space.zero()
+
+    forward_backward_pd(x, prox_f, prox_cc_g, lin_ops, grad_h, tau=0.5,
+                        sigma=[1.0], niter=10)
+
+    assert all_almost_equal(x, x_global_min, places=HIGH_ACCURACY)
+
+
+def test_forward_backward_with_lin_ops():
+    """Test for the forward-backward solver with linear operatros.
+
+    The test is done by minimizing ||x - b||_2^2 + ||alpha * x||_2^2. The
+    general problem is of the form
+
+        ``min_x f(x) + sum_i g_i(L_i x) + h(x)``
+
+    and here we take f = 0, g = ||.||_2^2, L = alpha * IndentityOperator,
+    and h = ||. - b||_2^2.
+    """
+
+    space = odl.rn(10)
+    scaling = 0.1
+    b = example_element(space)
+
+    lin_op = scaling * odl.IdentityOperator(space)
+    lin_ops = [lin_op]
+    prox_cc_g = [odl.solvers.proximal_cconj_l2_squared(space)]
+    prox_f = odl.solvers.proximal_zero(space)
+    # Gradient of two-norm square
+    grad_h = 2.0 * (odl.IdentityOperator(space) - odl.ConstantOperator(b))
+
+    x = example_element(space)
+
+    # Explicit solution: x_hat = (I^T * I + (alpha*I)^T * (alpha*I))^-1 * (I*b)
```

Comment:
Further: depending on another solver (in the case CG) makes reading the test quite hard, and can also (at least theoretically) give us circular tests, i.e. both fail but we fail to find it because the tests depend on each other.